lib/lpc55-rng/src/lib.rs

```diff
@@ -0,0 +1,74 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#![no_std]
+#![no_main]
+
+use core::{cmp, mem};
+use drv_lpc55_syscon_api::{Peripheral, Syscon};
+use drv_rng_api::RngError;
+use lpc55_pac::{pmc, rng, PMC, RNG};
+use rand_core::{impls, Error, RngCore};
+
+/// The Lpc55Rng is a thin wrapper around the LPC55 hardware random number
+/// generator (HRNG).
+pub struct Lpc55Rng {
+    pub pmc: &'static pmc::RegisterBlock,
+    pub rng: &'static rng::RegisterBlock,
+}
+
+impl Lpc55Rng {
+    /// Create a new Lpc55Rng instance after powering on, enabling the clocks
+    /// and reseting the underlying HRNG.
+    pub fn new(syscon: &Syscon) -> Self {
+        let pmc = unsafe { &*PMC::ptr() };
+        pmc.pdruncfg0.modify(|_, w| w.pden_rng().poweredon());
+
+        syscon.enable_clock(Peripheral::Rng);
+        syscon.enter_reset(Peripheral::Rng);
+        syscon.leave_reset(Peripheral::Rng);
+
+        Lpc55Rng {
+            pmc,
+            rng: unsafe { &*RNG::ptr() },
+        }
+    }
+}
+
+impl RngCore for Lpc55Rng {
+    /// Get the next 4 bytes from the HRNG.
+    fn next_u32(&mut self) -> u32 {
+        impls::next_u32_via_fill(self)
+    }
+
+    /// Get the next 8 bytes from the HRNG.
+    fn next_u64(&mut self) -> u64 {
+        impls::next_u64_via_fill(self)
+    }
+
+    /// Fill the provided buffer with output from the HRNG.
+    fn fill_bytes(&mut self, bytes: &mut [u8]) {
+        self.try_fill_bytes(bytes).expect("fill_bytes")
+    }
+
+    /// Fill the provided buffer with output from the HRNG. If the HRNG
+    /// can't service the request an error is returned.
+    fn try_fill_bytes(&mut self, dst: &mut [u8]) -> Result<(), Error> {
+        let mut filled = 0;
+        while filled < dst.len() {
+            if self.pmc.pdruncfg0.read().pden_rng().bits() {
+                return Err(RngError::PoweredOff.into());
```

Comment:
So, this looks like the only way this operation can fail. Given that the RNG is powered on in `new`, finding it powered off would be a surprising program bug. Perhaps this should be a panic? Otherwise I expect all callers to wind up unwrapping this at some point... could be wrong though.

Reply:
I went back and forth on whether to panic or return the error here a few times. Given the change requested in https://github.com/oxidecomputer/hubris/pull/1820#discussion_r1772082662 `new` will now take ownership of instances of the PMC & RNG types which makes it clear that we don't intend for them to change behind our backs. Returning the error was part of an attempt to be resilient to this, but it's so far from the common case that it's not worth making things less ergonomic.

Reply:
addressed by: [c29dffa](https://github.com/oxidecomputer/hubris/pull/1820/commits/c29dffad19ac1a26bcfe4a058c66efe5e016bc83)